### PR TITLE
[WIP] Add OrangePi R1 SPI support

### PR DESCRIPTION
**Pull request by Lankaster**
Flash partitions:
* u-boot: 0, 0xe0000 (896k - 917 504)
* u-boot-env: 0xe0000, 0x10000 (64k - 65 535)
* dtb: 0xf0000, 0x10000 (64k - 65 535)
* firmware: 0x100000, 0xF00000 (15360k - 15 728 640)

Notes:
* U-Boot can be installed without SD-Card via sunxi-fel

U-Boot log:
U-Boot SPL 2020.04 (Jul 05 2020 - 22:07:38 +0000)
DRAM: 256 MiB
Trying to boot from sunxi SPI

U-Boot 2020.04 (Jul 05 2020 - 22:07:38 +0000) Allwinner Technology

CPU:   Allwinner H3 (SUN8I 1680)
Model: Xunlong Orange Pi R1
DRAM:  256 MiB
MMC:   mmc@1c0f000: 0, mmc@1c10000: 1
Loading Environment from FAT... MMC: no card present
In:    serial
Out:   serial
Err:   serial
Net:   phy interface0
eth0: ethernet@1c30000
starting USB...
Bus usb@1c1a000: USB EHCI 1.00
Bus usb@1c1a400: USB OHCI 1.0
Bus usb@1c1b000: USB EHCI 1.00
scanning bus usb@1c1a000 for devices... 1 USB Device(s) found
scanning bus usb@1c1a400 for devices... 1 USB Device(s) found
scanning bus usb@1c1b000 for devices... 1 USB Device(s) found
       scanning usb for storage devices... 0 Storage Device(s) found
Hit any key to stop autoboot:  0
=>
=>
=> sf probe
SF: Detected mx25l12805d with page size 256 Bytes, erase size 64 KiB, total 16 MiB
=>

Based on:
* https://git.openwrt.org/?p=openwrt/staging/hauke.git;a=shortlog;h=refs/heads/sunxi-spi
* https://forum.openwrt.org/t/xunlong-orange-pi-r1-boot-from-flash/57700/21
* https://github.com/openwrt/openwrt/blob/f003d732d726758f11f0c308758637c434c37c1a/target/linux/sunxi/patches-4.19/400-arm64-allwinner-a64-sopine-Add-Sopine-flash-partitio.patch
* https://github.com/StephanvanSchaik/u-boot/tree/sunxi-spi
* https://lists.denx.de/pipermail/u-boot/2017-February/280966.html

### CODE REVIEW

File: package/boot/uboot-sunxi/patches/300-enable-orangepir1-spi.patch [hunk 1]
```diff
@@ -0,0 +1,18 @@
...
+ };
+ 
+@@ -72,6 +73,7 @@
+ 	status = "okay";
+ 
+ 	flash@0 {
++		u-boot,dm-pre-reloc;
+ 		compatible = "mxicy,mx25l12805d", "jedec,spi-nor";
+ 	};
+ };
```
Comment (adrianschmutzler) [inline package/boot/uboot-sunxi/patches/300-enable-orangepir1-spi.patch:18]: Please add newline at EOF.

File: target/linux/sunxi/image/Makefile [hunk 2]
```diff
@@ -30,6 +30,14 @@ define Build/sunxi-sdcard
 	rm -f $@.boot
 endef
 
+define Build/append-uboot
+	dd if=$(STAGING_DIR_IMAGE)/$(DEVICE_NAME)-u-boot-with-spl.bin >> $@
+endef
+
+define Build/add-dtb
```
Comment (adrianschmutzler) [inline target/linux/sunxi/image/Makefile:37]: Please move those blocks according to alphabetic sorting.

File: target/linux/sunxi/image/Makefile [hunk 3]
```diff
@@ -42,6 +50,20 @@ define Device/Default
...
+  BLOCKSIZE := 64k
+  IMAGE_SIZE := 15360k
+  IMAGE/spi-factory.img := append-uboot | check-size 896k | pad-to 960k | add-dtb | pad-to 1024k | append-kernel | append-rootfs | pad-rootfs | append-metadata
+  IMAGE/sysupgrade.bin := append-kernel | append-rootfs | pad-rootfs | append-metadata | check-size $$$$(IMAGE_SIZE)
+  DEVICE_TITLE:=Xunlong Orange Pi R1 boot from SPI
+  DEVICE_PACKAGES:=kmod-rtc-sunxi kmod-usb-net kmod-usb-net-rtl8152
+  SUPPORTED_DEVICES:=xunlong,orangepi-r1-spi
+  SUNXI_DTS:=sun8i-h2-plus-orangepi-r1
+endef
+TARGET_DEVICES += xunlong_orengepi-r1-spi
```
Comment (adrianschmutzler) [inline target/linux/sunxi/image/Makefile:65]: This needs to be put in one of the subtargets.

File: package/boot/uboot-sunxi/patches/301-add-orangepi-spi-partitions.patch [hunk 4]
```diff
@@ -0,0 +1,16 @@
...
++CONFIG_SPI=y
++CONFIG_SUN4I_SPI=y
++CONFIG_CMD_SF=y
++CONFIG_CMD_SPI=y
++CONFIG_DM_SPI=y
++CONFIG_DM_SPI_FLASH=y
++CONFIG_SPI_FLASH=y
++CONFIG_SPI_FLASH_WINBOND=y
++CONFIG_SPI_FLASH_MACRONIX=y
+\ No newline at end of file
```
Comment (adrianschmutzler) [inline package/boot/uboot-sunxi/patches/301-add-orangepi-spi-partitions.patch:16]: Please add a newline at the end of the patched file (not the patch itself).

**Comment by adrianschmutzler** [comment]
This has formal issues; from a quick look, I see at least missing commit title prefix and missing Signed-off-by.
Please have another look at https://openwrt.org/submitting-patches

**Comment by Lankaster** [comment]
@adrianschmutzler Thank you for realy fast review. I'll test SPI Image next weeks and prepare MR: commit title prefix, squash commits, commit message etc.

**Comment by adrianschmutzler** [comment]
@Lankaster FYI: I won't merge device support on sunxi, just commented on the obvious stuff here.
Note that if you are done, you are also supposed to squash all changes into a single proper commit (git rebase -i ...) and force-push that one to this PR again (git push -f ....).

**Comment by 981213** [comment]
Hi! Could you try building a FIT kernel instead of supplying kernel/dtb separately?
Add the following into your image recipe:
```
  KERNEL_NAME := Image
  KERNEL = kernel-bin | lzma | \
	fit lzma $$(KDIR)/image-$$(firstword $$(DEVICE_DTS)).dtb
  KERNEL_INITRAMFS = kernel-bin | lzma | \
	fit lzma $$(KDIR)/image-$$(firstword $$(DEVICE_DTS)).dtb
  FILESYSTEMS := squashfs
  DEVICE_DTS_DIR := $(DTS_DIR)
  DEVICE_DTS:=sun8i-h2-plus-orangepi-r1
```
and drop dtb related stuff from factory image recipe, partition table and u-boot-env

**Comment by Lankaster** [comment]
@981213 Wow, it looks great. I'll try it.

**Comment by Lankaster** [comment]
@981213 It doesn't work. I'll try to increase CONFIG_SYS_BOOTM_LEN, but it doesn't help. Any suggestions ?

```
sf read $kernel_addr_r 0x80000 0x400000
device 0 offset 0x80000, size 0x400000
SF: 4194304 bytes @ 0x80000 Read: OK
=> setenv bootargs console=ttyS0,115200 earlyprintk root=/dev/mtdblock3 rootwait mtdparts=spi0.0:512k(uboot),-(firmware)
=> bootm $kernel_addr_r
## Loading kernel from FIT Image at 42000000 ...
   Using 'config@1' configuration
   Trying 'kernel@1' kernel subimage
     Description:  ARM OpenWrt Linux-5.4.52
     Type:         Kernel Image
     Compression:  lzma compressed
     Data Start:   0x420000e4
     Data Size:    3175376 Bytes = 3 MiB
     Architecture: ARM
     OS:           Linux
     Load Address: 0x40008000
     Entry Point:  0x40008000
     Hash algo:    crc32
     Hash value:   98c9ff24
     Hash algo:    sha1
     Hash value:   26e09234fbd6948178dc5c9a6b55c175046e33d4
   Verifying Hash Integrity ... crc32+ sha1+ OK
## Loading fdt from FIT Image at 42000000 ...
   Using 'config@1' configuration                                                         +-----------------------------+
   Trying 'fdt@1' fdt subimage                                                            |                             |
     Description:  ARM OpenWrt xunlong_orangepi-r1-spi device tree blob                   |  Cannot open /dev/ttyUSB0!  |
     Type:         Flat Device Tree                                                       |                             |
     Compression:  uncompressed                                                           +-----------------------------+
     Data Start:   0x423075f8
     Data Size:    20798 Bytes = 20.3 KiB
     Architecture: ARM
     Hash algo:    crc32
     Hash value:   418f3a32
     Hash algo:    sha1
     Hash value:   8b8a048c3bbf992d0eea84afc079737e4f9e64b7
   Verifying Hash Integrity ... crc32+ sha1+ OK
   Booting using the fdt blob at 0x423075f8
EHCI failed to shut down host controller.
   Uncompressing Kernel Image
Image too large: increase CONFIG_SYS_BOOTM_LEN
Must RESET board to recover
resetting ...
```

File: package/boot/uboot-sunxi/patches/302-increase-sys-bootm-len.patch [hunk 5]
```diff
@@ -0,0 +1,11 @@
+--- a/include/configs/sunxi-common.h
++++ b/include/configs/sunxi-common.h
+@@ -35,6 +35,8 @@
+ #define CONFIG_SYS_BOOTM_LEN		(32 << 20)
+ #endif
+ 
++#define CONFIG_SYS_BOOTM_LEN            0x10000000
```
Comment (Lankaster) [inline package/boot/uboot-sunxi/patches/302-increase-sys-bootm-len.patch:7]: Try to increase up to 256Mb, but It doesn't work.

File: target/linux/sunxi/image/cortexa7.mk [hunk 6]
```diff
@@ -178,6 +178,28 @@ define Device/xunlong_orangepi-r1
...
+  KERNEL_INITRAMFS = kernel-bin | \
+    fit lzma $$(KDIR)/image-$$(firstword $$(DEVICE_DTS)).dtb
+  FILESYSTEMS := squashfs
+  DEVICE_DTS_DIR := $(DTS_DIR)
+  DEVICE_DTS:=sun8i-h2-plus-orangepi-r1
+  IMAGES := spi-factory.img sysupgrade.bin
+  BLOCKSIZE := 64k
+  IMAGE_SIZE := 15872k
+  IMAGE/spi-factory.img := append-uboot | check-size 512k | pad-to 512k | append-kernel | append-rootfs | pad-rootfs | append-metadata
+  IMAGE/sysupgrade.bin := append-kernel | append-rootfs | pad-rootfs | append-metadata | check-size $$$$(IMAGE_SIZE)
```
Comment (adschm) [inline target/linux/sunxi/image/cortexa7.mk:198]: check-size can be used without argument here.

File: target/linux/sunxi/image/cortexa7.mk [hunk 7]
```diff
@@ -178,6 +178,28 @@ define Device/xunlong_orangepi-r1
...
+  SUPPORTED_DEVICES += orangepi-r1-spi
```
Comment (adschm) [inline target/linux/sunxi/image/cortexa7.mk:199]: Is there a reason for adding this?

**Comment by adschm** [comment]
@Lankaster Ping. Do you plan to continue working on this or can we close it?

**Comment by Lankaster** [comment]
@adschm My first concept working, but implementation was not good enough. I think it make more sense to implement FIT support ( thank you @981213 !). Unfortunately last year i doesn't have a time to continue this work. As soon as I get time this year, I'll continue working on SPI support.
I think right now it make sense to close this issue.

**Pull request closed without merging**